src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py

```diff
@@ -88,9 +88,34 @@ def create_clone(self, pool, source_volname, source_subvolume, snapname):
 
         subvol_path = os.path.join(self.base_path, str(uuid.uuid4()).encode('utf-8'))
         try:
+            # source subvolume and snapshot attrs are used to create clone subvolume
+            # attributes of subvolume's content though, are synced during the cloning process
+            src_snappath = source_subvolume.snapshot_path(snapname)
+            st = source_subvolume.fs.stat(src_snappath)
+            uid = int(st.st_uid)
+            gid = int(st.st_gid)
+            mode = int(st.st_mode & ~stat.S_IFMT(st.st_mode))
+
+            # apply source subvolume size restrictions to the clone
+            try:
+                size = int(source_subvolume.fs.getxattr(source_subvolume.path, 'ceph.quota.max_bytes').decode('utf-8'))
```

Comment:
If source subvolume size restrictions are applied to the clone, won't the clone fail in following scenario.

1. Create subvolume of size 1GB (assume 1GB is written)
2. Take a subvolume snapshot
3. Delete 500 MB and resize the subvolume to 500 MB
4. Clone the subvolume snapshot

Reply:
Good point. I tried this out. As expected the clone status of the original snapshot remains in progress, and I see the following traceback,
```
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/fs_util.py", line 117, in copy_file
    written += fs.write(dst_fd, data[written:], -1)
  File "cephfs.pyx", line 1463, in cephfs.LibCephFS.write
cephfs.Error: error in write: Disk quota exceeded [Errno 122]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_job.py", line 44, in run
    self.async_job.execute_job(vol_job[0], vol_job[1], should_cancel=lambda: thread_id.should_cancel())
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 309, in execute_job
    clone(self.vc, volname, job[0].decode('utf-8'), job[1].decode('utf-8'), self.state_table, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 222, in clone
    start_clone_sm(volume_client, volname, index, groupname, subvolname, state_table, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 202, in start_clone_sm
    (next_state, finished) = handler(volume_client, volname, index, groupname, subvolname, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 159, in handle_clone_in_progress
    do_clone(volume_client, volname, groupname, subvolname, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 155, in do_clone
    bulk_copy(fs_handle, src_path, dst_path, should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 144, in bulk_copy
    cptree(source_path, dst_path)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/async_cloner.py", line 129, in cptree
    copy_file(fs_handle, d_full_src, d_full_dst, mo, cancel_check=should_cancel)
  File "/home/rraja/git/ceph/src/pybind/mgr/volumes/fs/fs_util.py", line 120, in copy_file
    raise VolumeException(-e.args[0], e.args[1])
TypeError: bad operand type for unary -: 'str'
```

Reply:
I expected the snapshot to preserve the xattr `ceph.quota.max_bytes` like it does for `ceph.dir.layout`, but it doesn't it. Not sure why.

Reply:
> Good point. I tried this out. As expected the `fs clone status` of the original snapshot's clone remains in progress, and I see the following traceback,

Clone status should be "failed" -- guess the async cloner hit an unhandled exception here.

So, the quota xattr is not preserved in the snapshot -- that seems like a bug.

Reply:
The xattr should be synced from the snapshot anyway...

Reply:
The quota xattr applies to the subvolume and the size occupied by its snapshots as well, I would assume.

IOW, creating content -> taking a snapshot -> deleting content, would allow a user to exceed the quota restriction on the subvolume, and store content in the snapshots instead.

Considering this, I think the manner of solving this would be to store the quota at the time of snapshot in the metadata file, like we intend to do for snapshot df content.

Reply:
> The quota xattr applies to the subvolume and the size occupied by its snapshots as well, I would assume.
> 
> IOW, creating content -> taking a snapshot -> deleting content, would allow a user to exceed the quota restriction on the subvolume, and store content in the snapshots instead.

Tested the above, and found out that snapshot COW blocks are not accounted for in the directory quota. Other results present in https://tracker.ceph.com/issues/46278 

> 
> Considering this, I think the manner of solving this would be to store the quota at the time of snapshot in the metadata file, like we intend to do for snapshot df content.

With the observations in the tracker, this can be solved fixing the xattr being available in the snapshots, or for subvolumes storing this in the meta as mentioned above.

Reply:
> > Good point. I tried this out. As expected the `fs clone status` of the original snapshot's clone remains in progress, and I see the following traceback,
> 
> Clone status should be "failed" -- guess the async cloner hit an unhandled exception here.
> 
> So, the quota xattr is not preserved in the snapshot -- that seems like a bug.

https://tracker.ceph.com/issues/46360
Also posted fix

Reply:
Addressed picking up quota from the snapshot as related tracker https://tracker.ceph.com/issues/46278 is now fixed.